pkg/controllers/vdb/promotesandboxtomain_reconciler.go

```diff
@@ -0,0 +1,93 @@
+	"context"
+
+	"github.com/go-logr/logr"
+	vapi "github.com/vertica/vertica-kubernetes/api/v1"
+	"github.com/vertica/vertica-kubernetes/pkg/controllers"
+	"github.com/vertica/vertica-kubernetes/pkg/events"
+	"github.com/vertica/vertica-kubernetes/pkg/vadmin"
+	"github.com/vertica/vertica-kubernetes/pkg/vadmin/opts/promotesandboxtomain"
+
+	corev1 "k8s.io/api/core/v1"
+	ctrl "sigs.k8s.io/controller-runtime"
+)
+
+// PromoteSandboxSubclusterToMainReconciler will convert local sandbox to main cluster
+type PromoteSandboxSubclusterToMainReconciler struct {
+	VRec        *VerticaDBReconciler
+	Log         logr.Logger
+	Vdb         *vapi.VerticaDB // Vdb is the CRD we are acting on
+	PFacts      *PodFacts
+	InitiatorIP string // The IP of the pod that we run vclusterOps from
+	Dispatcher  vadmin.Dispatcher
+}
+
+// MakePromoteSandboxSubclusterToMainReconciler will build a promoteSandboxSubclusterToMainReconciler object
+func MakePromoteSandboxSubclusterToMainReconciler(vdbrecon *VerticaDBReconciler, log logr.Logger, vdb *vapi.VerticaDB,
+	pfacts *PodFacts, dispatcher vadmin.Dispatcher) controllers.ReconcileActor {
+	return &PromoteSandboxSubclusterToMainReconciler{
+		VRec:       vdbrecon,
+		Log:        log.WithName("promoteSandboxSubclusterToMainReconciler"),
+		Vdb:        vdb,
+		PFacts:     pfacts,
+		Dispatcher: dispatcher,
+	}
+}
+
+func (s *PromoteSandboxSubclusterToMainReconciler) Reconcile(ctx context.Context, _ *ctrl.Request) (ctrl.Result, error) {
+	// no-op for ScheduleOnly init policy or enterprise db or no sandboxes
+	if s.Vdb.Spec.InitPolicy == vapi.CommunalInitPolicyScheduleOnly ||
+		!s.Vdb.IsEON() || len(s.Vdb.Spec.Sandboxes) == 0 {
+		return ctrl.Result{}, nil
+	}
+
+	// We need to collect pod facts for finding qualified subclusters
+	if err := s.PFacts.Collect(ctx, s.Vdb); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	return s.promoteSandboxToMain(ctx)
+}
+
+// promoteSandboxToMain call the vclusterOps API to convert local sandbox to main cluster
+func (s *PromoteSandboxSubclusterToMainReconciler) promoteSandboxToMain(ctx context.Context) (ctrl.Result, error) {
+	initiator, ok := s.PFacts.findFirstPodSorted(func(v *PodFact) bool {
+		return v.isPrimary && v.upNode
+	})
+	if !ok {
+		s.Log.Info("No Up nodes found. Requeue reconciliation.")
+		return ctrl.Result{Requeue: true}, nil
+	}
+	// pick first sandbox to promote
+	sandboxName := s.Vdb.Spec.Sandboxes[0].Name
+	s.VRec.Eventf(s.Vdb, corev1.EventTypeNormal, events.PromoteSandboxToMainStart,
+		"Starting promote sandbox %q to main", sandboxName)
+
+	err := s.Dispatcher.PromoteSandboxToMain(ctx,
+		promotesandboxtomain.WithInitiator(initiator.podIP),
+		promotesandboxtomain.WithSandbox(sandboxName),
+	)
+	if err != nil {
+		s.VRec.Eventf(s.Vdb, corev1.EventTypeWarning, events.PromoteSandboxToMainFailed,
+			"Failed to promote sandbox %q to main", sandboxName)
+		return ctrl.Result{}, err
+	}
+	s.VRec.Eventf(s.Vdb, corev1.EventTypeNormal, events.PromoteSandboxToSucceeded,
```

Comment:
After promoting sandbox to main, we need to update vdb to remove sandbox of both spec and status. Moreover, we need to update "SandboxPrimarySubcluster" to "PrimarySubcluster" in vdb.sepc.subclusters.